salt/states/pkg.py

```diff
@@ -872,6 +880,27 @@ def _nested_output(obj):
     ret = nested.output(obj).rstrip()
     return ret
 
+def _resolve_capabilities(pkgs, refresh=False, **kwargs):
+    '''
+    Resolve capabilities in ``pkgs`` and exchange them with real package
+    names, when the result is distinct.
+    This feature can be turned on while setting the paramter
+    ``resolve_capabilities`` to True.
+
+    Return the input dictionary with replaced capability names and as
+    second return value a bool which indicate if a refresh was run.
+
+    In case of ``resolve_capabilities`` is False (disabled) or not
+    supported by the implementation the input is returned unchanged.
+    '''
+    was_refreshed = False
+    if not pkgs or 'pkg.resolve_capabilities' not in __salt__:
+        return (pkgs, was_refreshed)
+
+    ret = __salt__['pkg.resolve_capabilities'](pkgs, refresh=refresh, **kwargs)
+    was_refreshed = refresh
+    return (ret, was_refreshed)
```

Comment:
I would still return here:
```python
return ret, not refresh
```
This removes strange check every time if it WAS refreshed, THEREFORE do not refresh (or vice versa). Just say from this function it needs to be refreshed or not. So you would just:

```python
if pkgs:
    pkgs, refresh = _resolve_capabilities(pkgs, refresh=refresh, **kwargs)
```
Extra parenthesis in `return` aren't needed, BTW. Same when you're assigning such return later (you're doing it below everywhere).

Reply:
I like the idea, but 

    return ret, not refresh

is wrong. The goal is to refresh only 1 time. But using

    return ret, False

should work.